src/modules/rlm_imap/rlm_imap.c

```diff
@@ -0,0 +1,264 @@
+static fr_dict_attr_t 	const 		*attr_user_name;
+
+extern fr_dict_autoload_t rlm_imap_dict[];
+fr_dict_autoload_t rlm_imap_dict[] = {
+	{ .out = &dict_radius, .proto = "radius" },
+	{ .out = &dict_freeradius, .proto = "freeradius" },
+	{ NULL }
+};
+
+extern fr_dict_attr_autoload_t rlm_imap_dict_attr[];
+fr_dict_attr_autoload_t rlm_imap_dict_attr[] = {
+	{ .out = &attr_auth_type, .name = "Auth-Type", .type = FR_TYPE_UINT32, .dict = &dict_freeradius },
+	{ .out = &attr_user_name, .name = "User-Name", .type = FR_TYPE_STRING, .dict = &dict_radius },
+	{ .out = &attr_user_password, .name = "User-Password", .type = FR_TYPE_STRING, .dict = &dict_radius },
+	{ NULL },
+};
+
+typedef struct {
+	fr_dict_enum_t        		*auth_type;
+	char const        		*name;
+	char const			*imap_uri;	//!<URL of imap server
+	fr_time_delta_t 		timeout;	//!<Timeout for connection and server response
+	fr_curl_tls_t			tls;
+} rlm_imap_t;
+
+typedef struct {
+	rlm_imap_t const    		*inst;        //!< Instance of rlm_imap.
+	fr_curl_handle_t    		*mhandle;    //!< Thread specific multi handle.  Serves as the dispatch and coralling structure for imap requests.
+}rlm_imap_thread_t;
+
+typedef struct {
+	rlm_imap_t const    		*instance;    //!< This instance of rlm_imap.
+	REQUEST           		*request;    //!< Current request.
+} rlm_imap_request_t;
+
+static const CONF_PARSER module_config[] = {
+	{ FR_CONF_OFFSET("imap_uri", FR_TYPE_STRING, rlm_imap_t, imap_uri) },
+	{ FR_CONF_OFFSET("timeout",FR_TYPE_TIME_DELTA, rlm_imap_t, timeout) },
+	{ FR_CONF_OFFSET("tls", FR_TYPE_SUBSECTION, rlm_imap_t, tls), .subcs = (void const *) fr_curl_tls_config },//!<loading the tls values
+	CONF_PARSER_TERMINATOR
+};
+
+/*
+ * Check to see that we have a username and password
+ * print out the username
+ * if a value is missing return NOOP
+ */
+static rlm_rcode_t mod_authorize(void *instance, UNUSED void *thread, REQUEST *request)
+{
+	VALUE_PAIR 			*vp;
+	rlm_imap_t			*inst	= instance;
+
+	vp = fr_pair_find_by_da(request->packet->vps, attr_user_name, TAG_ANY);
+	if (vp == NULL) return RLM_MODULE_NOOP;
+	RDEBUG2("Authorizing: %pP", vp);
+	vp = fr_pair_find_by_da(request->packet->vps, attr_user_password, TAG_ANY);
+	if (vp == NULL) return RLM_MODULE_NOOP;
+	/*
+	 *Check to see if there is already a section that will be working on the request
+	 */
+	if (!module_section_type_set(request, attr_auth_type, inst->auth_type)) return RLM_MODULE_NOOP;
+
+	return RLM_MODULE_OK;
+}
+/*
+ *Called when the IMAP server responds
+ *It checks if the response was CURLE_OK
+ *If it wasn't we returns REJECT, if it was we returns OK
+*/
+static rlm_rcode_t CC_HINT(nonnull) mod_authenticate_resume(void *instance, UNUSED void *thread, REQUEST *request, void *rctx)
+{
+	fr_curl_io_request_t     	*randle   = rctx;
+	rlm_imap_t			*inst 	= instance;
+	fr_curl_tls_t			*tls;
+	long 				curl_out;
+
+	tls = &inst->tls;
+
+	if (randle->result != CURLE_OK) {
+		talloc_free(randle);
```

Comment:
Not sure, but maybe we could see a `double-free` in https://github.com/FreeRADIUS/freeradius-server/pull/3469/files#diff-a1f19cf69ca098dbd021c5543dae1a47R240 due to be freed here at `mod_authenticate_resume()`?